File: cpp/samples/06_event_listener.cpp
/**
 * Sample 06 — Event Listener
 *
 * Demonstrates:
 *   - RedfishEventListener construction
 *   - listener.use_context(ctx)
 *   - listener.on_event() global callback
 *   - listener.start() / listener.stop()
 *   - Auto-subscribe with listener.listen_url() as destination
 *   - Waiting for events then cleaning up
 *
 * Usage:
 *   ./06_event_listener [host] [port] [listen-port] [wait-seconds]
 */
#include "redfish_sdk/client.hpp"
#include "redfish_sdk/errors.hpp"
#include "redfish_sdk/event_listener.hpp"
#include <atomic>
#include <chrono>
#include <iostream>
#include <thread>

int main(int argc, char* argv[]) {
    std::string host        = (argc > 1) ? argv[1] : "127.0.0.1";
    int         port        = (argc > 2) ? std::stoi(argv[2]) : 8000;
    uint16_t    listen_port = (argc > 3) ? static_cast<uint16_t>(std::stoi(argv[3])) : 9090;
    int         wait_sec    = (argc > 4) ? std::stoi(argv[4]) : 10;

    redfish::Credentials    creds{"admin", "admin"};
    redfish::ConnectionConfig config;
    config.verify_tls = false;  // disable cert verification for simulator (self-signed)
    for (int i = 1; i < argc; ++i)
        if (std::string(argv[i]) == "--no-tls") config.use_tls = false;  // plain HTTP

    try {
        auto ctx = redfish::connect(host, port, creds,
                                    redfish::AuthMode::SESSION, config);

        // ── Build listener ─────────────────────────────────────────────
        redfish::RedfishEventListener listener{listen_port, "0.0.0.0",
                                               "RSDK-Sample-06"};
        listener.use_context(*ctx);

        std::atomic<int> received{0};

        listener.on_event([&](const redfish::RedfishEvent& ev) {
            ++received;
            std::cout << "  [EVENT] type=" << ev.event_type
                      << "  id=" << ev.message_id << "\n"
                      << "          msg=" << ev.message << "\n"
                      << "          sev=" << ev.severity << "\n";
        });

        listener.start();
        std::string url = listener.listen_url();
        std::cout << "✓ Listener running at " << url << "\n";

        // ── Subscribe ─────────────────────────────────────────────────
        // Use 127.0.0.1 as the destination so the simulator can reach us.
        std::string dest = "http://127.0.0.1:" + std::to_string(listen_port)
                           + "/events";
        std::cout << "Subscribing → " << dest << "\n";

        auto& ev_svc = ctx->events();
        auto  sub    = ev_svc.subscribe(dest,
                                         {"Alert", "ResourceUpdated"},
                                         {}, {}, {}, "",
                                         "RSDK-Sample-06");
        std::string sub_uri;
        if (sub.success) {
            sub_uri = sub.location();
            if (sub_uri.empty() && sub.body.is_object())
                sub_uri = sub.body.value("@odata.id", "");
            std::cout << "  ✓ Subscribed: " << sub_uri << "\n";
        } else {
            std::cout << "  ✗ Subscribe failed HTTP " << sub.status_code << "\n";
        }

        // ── Trigger a test event ──────────────────────────────────────
        std::cout << "Submitting test event …\n";
        ev_svc.submit_test_event();

        // ── Wait ──────────────────────────────────────────────────────
        std::cout << "Waiting " << wait_sec << "s for events …\n";
        std::this_thread::sleep_for(std::chrono::seconds(wait_sec));

        std::cout << "\nReceived " << received.load() << " event(s) in "
                  << wait_sec << "s\n";

        // ── Buffered events via GET ───────────────────────────────────
        auto buffered = listener.get_buffered_events();
        std::cout << "Buffered events in ring buffer: " << buffered.size() << "\n";

        // ── IP stats ──────────────────────────────────────────────────
        auto stats = listener.get_ip_stats();
        for (const auto& [ip, count] : stats)
            std::cout << "  IP " << ip << " → " << count << " event(s)\n";

        // ── Cleanup ───────────────────────────────────────────────────
        if (!sub_uri.empty()) {
            ev_svc.delete_subscription(sub_uri);
            std::cout << "✓ Subscription deleted\n";
        }
        listener.stop();
        std::cout << "✓ Listener stopped\n";

        std::cout << "\n✓ Event listener sample complete\n";
        return 0;
    } catch (const redfish::RedfishError& e) {
        std::cerr << "[ERROR] " << e.what() << "\n";
    }
    return 1;
}

File: python/samples/06_event_listener.py
# SPDX-License-Identifier: MIT
# Copyright (c) Microsoft Corporation. All rights reserved.

#!/usr/bin/env python3
"""Sample 06 — Start embedded RedfishEventListener and auto-subscribe.

Demonstrates:
  - RedfishEventListener instantiation
  - listener.use_context(ctx)
  - listener.on_event() global callback
  - listener.start() / listener.stop()
  - Auto-subscribe with listener.listen_url as destination
  - Waiting for test events with a timeout

The listener binds on the local machine; the subscription destination is
set to the listener URL so the BMC (or simulator) POSTs events here.

Usage:
    python 06_event_listener.py [--host HOST] [--port PORT]
                                [--listen-host LISTEN_HOST]
                                [--listen-port LISTEN_PORT]
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import redfish_sdk
from redfish_sdk import AuthMode, ConnectionConfig, Credentials, RedfishEventListener
from redfish_sdk.services.event_service import RedfishEvent
from redfish_sdk.errors import RedfishConnectionError


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Sample 06 — Event Listener")
    p.add_argument("--host", default="127.0.0.1")
    p.add_argument("--port", type=int, default=8000)
    p.add_argument("--user", default="admin")
    p.add_argument("--password", default="password")
    p.add_argument("--no-tls-verify", action="store_true")
    p.add_argument("--no-tls", action="store_true", help="Use plain HTTP instead of HTTPS (for simulators without SSL)")
    p.add_argument(
        "--listen-host",
        default="0.0.0.0",
        help="Interface the listener binds to",
    )
    p.add_argument(
        "--listen-port",
        type=int,
        default=9090,
        help="Port the listener binds to",
    )
    p.add_argument(
        "--wait",
        type=float,
        default=10.0,
        help="Seconds to wait for events before exiting",
    )
    return p.parse_args()


async def main() -> None:
    args = parse_args()

    creds = Credentials(username=args.user, password=args.password)
    config = ConnectionConfig(verify_tls=not args.no_tls_verify, use_tls=not args.no_tls)

    try:
        ctx = await redfish_sdk.connect_async(
            host=args.host,
            port=args.port,
            credentials=creds,
            auth_mode=AuthMode.SESSION,
            config=config,
        )
    except RedfishConnectionError as exc:
        print(f"[ERROR] {exc}")
        sys.exit(1)

    async with ctx:
        # ── Build and start listener ────────────────────────────────────
        listener = RedfishEventListener(
            port=args.listen_port,
            host=args.listen_host,
        )
        listener.use_context(ctx)

        received: list[RedfishEvent] = []

        @listener.on_event
        async def on_any_event(event: RedfishEvent) -> None:
            received.append(event)
            print(f"  [EVENT] type={event.event_type!r}  id={event.message_id!r}")
            print(f"          msg ={event.message!r}")
            print(f"          sev ={event.severity!r}")

        listener.start()
        print(f"✓ Listener running at {listener.listen_url}")

        # ── Subscribe ────────────────────────────────────────────────────
        # Use 127.0.0.1 as destination so simulator can reach us locally
        dest = f"http://127.0.0.1:{args.listen_port}/events"
        print(f"Subscribing → {dest}")

        sub_resp = await ctx.events.subscribe_async(
            destination=dest,
            event_types=["Alert", "ResourceUpdated", "StatusChange"],
            context="RSDK-Sample-06",
        )

        sub_uri: str | None = None
        if sub_resp.success:
            sub_uri = sub_resp.headers.get("Location") or sub_resp.body.get("@odata.id")
            print(f"  ✓ Subscribed — URI: {sub_uri}")
        else:
            print(f"  ✗ Subscribe failed HTTP {sub_resp.status_code}; still waiting for events")

        # ── Trigger a test event ─────────────────────────────────────────
        print("Submitting test event …")
        await ctx.events.submit_test_event_async()

        # ── Wait for events ──────────────────────────────────────────────
        print(f"Waiting up to {args.wait}s for incoming events …")
        await asyncio.sleep(args.wait)

        print(f"\nReceived {len(received)} event(s) in {args.wait}s")

        # ── Cleanup ──────────────────────────────────────────────────────
        if sub_uri:
            await ctx.events.delete_subscription_async(sub_uri)
            print("✓ Subscription deleted")

        listener.stop()
        print("✓ Listener stopped")

        print("\n✓ Event listener sample complete")


if __name__ == "__main__":
    asyncio.run(main())
